Versioning and Stability for OpenTelemetry Clients (#143)
open-telemetry/oteps: text/0143-versioning-and-stability.md

# OpenTelemetry versioning and stability

> OpenTelemetry is a large project with strict compatibility requirements. This proposal defines the stability gurantees offered by OpenTelemetry, along with a versioning and lifecycle proposal which defines how we meet those requirements.

Reviewer: ```suggestion
OpenTelemetry is a large project with strict compatibility requirements. This proposal defines the stability guarantees offered by OpenTelemetry, along with a versioning and lifecycle proposal which defines how we meet those requirements.
```

> Language implementations are expected to follow this proposal exactly, unless a langague or package manager convention interferes significantly. Implementations must take this corss-language proposal, and produce a language-specific proposal which details how these requirements will be met.

Reviewer 2: ```suggestion
Language implementations are expected to follow this proposal exactly, unless a langague or package manager convention interferes significantly. Implementations must take this cross-language proposal, and produce a language-specific proposal which details how these requirements will be met.
```

## Design goals

**Provide end users with coherent, tested releases.**  

> We do not want users to run untested combinations of core packages. Users must be able to say “I am using OpenTelemetry v1.2.0 for Python” and have a shared understanding as to what that means.

Reviewer 2: TBH, I'm not sure what this means, even with this proposal. Is it the API version of a library? Is it an SDK version running in an application? Without that context, I think this statement will always be ambiguous.

Author: good point. Would saying that "the release number is the *highest supported* version of the SDK the API" clarify this? That is the intent, at least.

Reviewer 2: I don't grok that. Maybe provide the whole sentence?  

I think if you just put the word "API" or "SDK" in between "OpenTelemetry" and the version, it makes sense. 

The more I think about this, the more I don't understand what you're getting at, TBH. :)

Author: Ok, well maybe this is just a distraction and I should delete it. My main point is that all core components should be released together: Don't release a new version of the API without SDK and contrib support; release everything together. But this is a poor way of wording it.

Reviewer 3: I also am curious about the thought behind this.  If one is saying "I am using OpenTelemetry v1.2.0 for Python", that would most likely correlate to a tagged release?  While I agree that a version number <-> code understanding is important, I don't know whether our version numbering is prudent in this example.

Author: Yeah, honestly what I was trying to say isn't really clear or necessary, so I'll remove this.

> **Ensure that end users stay up to date with the latest release.**  

Reviewer 4: Do we have any requirements that it must also be possible for users to only upgrade some, but not all, of OpenTelemetry dependencies? In other words, are languages allowed to have a requirement that versions are aligned? API 1.1 is not guaranteed to work with SDK 1.2, but API 1.2 and SDK 1.2 are guaranteed to work, and support any published instrumentation / public API user out there due to backcompat.

I think in Java it's pretty conventional to require version alignment, and is pretty easy to achieve with BOMs. I think I've seen similar with JS (at least with React), don't know more broadly. Requiring version alignment has the benefit of allowing internal packages to be shared among artifacts within a repository, which I think is useful and sometimes even performance-boosting (e.g., if possible we would only want to introduce a single ThreadLocal string-building buffer in the app, this is only possible if we can share that among all the artifacts as part of an internal package). So I'm hoping that this is an allowed restriction, though can live without it too.

Reviewer 5: What if my application uses several libraries that used Otel API to instrument themselves? These libraries may use different minor versions of API and my app still should work.

Reviewer 4: Yeah - since we "Ensure that end users stay up to date with the latest release.", then it means that this will work no problem if the user uses the latest version of the API or SDK (strictly speaking, it only has to be the version of the latest version specified by a transitive dependency). We know the latest API is compatible with all the libraries (it must be per semver), and we know the lastest API is compatible with the latest SDK. As long as versions are aligned to a new version of things, there's no breakage.

Author: @anuraaga I completely agree. I would really prefer that the versions remain aligned, in terms of what libraries are ultimately loaded. Possibly, the API and the SDK need to have a separate version – but I would avoid that if I could. And even in that case, I agree that the user should always load the latest version of the API. Stable packages within the SDK definitely should have the same version number, which matches a tag on the repository. 

We don't want users updating some core packages, but not others. This would result in an untested configuration. I strongly prefer that we only have users running coherent, tested releases. No mix-n-match, please.

> We want all users to  stay up to date with the latest version of OpenTelemetry. We do not want to create hard breaks in support, of any kind, which leave users stranded on older versions. It must always be possible to upgrade to the latest minor version of OpenTelemetry, without creating a compilation error.

Author: Good point. We do mean ABI compatibility, but it does not need to be listed here

**Never create a dependency conflict between packages which rely on different versions of OpenTelemetry. Avoid breaking all stable public APIs.**  

> Backwards compatibility is a strict requirement. Instrumentation APIs cannot create a version conflict, ever. Otherwise, OpenTelemetry cannot be embedded in widely shared libraries, such as web frameworks. Theoretically, APIs can be deprecated and eventually removed, but this is a process measured in years.

Author: It means:
* ABI compatibility.
* No dependencies may be added which may create a transitive dependency conflict.

> Backwards compatibility is a strict requirement. Instrumentation APIs cannot create a version conflict, ever. Otherwise, OpenTelemetry cannot be embedded in widely shared libraries, such as web frameworks. Theoretically, APIs can be deprecated and eventually removed, but this is a process measured in years.

Reviewer 4: I think we can clarify that "any published library that uses OpenTelemetry must work without errors with the latest version of OpenTelemetry with the same major version" or something to the effect, it repeats a bit of the previous bullet, but is important to have in the context of libraries. Naturally, if a library requires features in 1.5, it can't work with 1.1, and that's fine

> Backwards compatibility is a strict requirement. Instrumentation APIs cannot create a version conflict, ever. Otherwise, OpenTelemetry cannot be embedded in widely shared libraries, such as web frameworks. Theoretically, APIs can be deprecated and eventually removed, but this is a process measured in years.

Author: Ok I added a clarification. There is an additional requirement you did not mention, which I want to point out: transitive dependencies must also not create a conflict. The OpenTelemetry API cannot depend on "foo" if there is any chance that any library or application may require an incompatible version of "foo."

> Provide maintainers a clear process for developing new, experimental APIs alongside stable APIs. DIfferent packages within the same release will have different levels of stability.

Reviewer 5: ```suggestion
Provide maintainers a clear process for developing new, experimental APIs alongside stable APIs. DIfferent packages within the same release MAY have different levels of stability.
```

**Allow for multiple levels of package stability within the same release.**  

> Provide maintainers a clear process for developing new, experimental APIs alongside stable APIs. DIfferent packages within the same release will have different levels of stability.

Reviewer 3: That's how I'm reading this and how it's seemed like @tedsuo has pitched this, but I'll defer to him on that solid definition.  Might be worth including that in the OTEP too.

> Provide maintainers a clear process for developing new, experimental APIs alongside stable APIs. DIfferent packages within the same release will have different levels of stability.

Author: Yes, everything except for tracing is still experimental.

That means that an implementation wishing to release stable tracing today must ensure that metrics are factored out in such a way that breaking changes to the metrics API would not destabilize the trace API packages.

> ## Relevant architecure

Reviewer: ```suggestion
## Relevant architecture
```

> OpenTelemetry is split into signals. Each signal provides a specialized form of observability.

Author: Makes sense, I can add more detail. Be should use this term more frequently. :)

> OpenTelemetry is split into signals. Each signal provides a specialized form of observability.

Reviewer 6: We already indirectly mention this in the [glossary](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/glossary.md#instrumented-library) - but we should actually add it there for easy quoting ;)

> At the highest architectural level, OpenTelemetry is organized into signals. Each signal provides a specialized form of observability. For example, tracing, metrics, and baggage are three seperate signals. Signals share a common subsystem – context propagation – but they function independently from each other.

Reviewer 5: ```suggestion
At the highest architectural level, OpenTelemetry is organized into signals. Each signal provides a specialized form of observability. For example, tracing, metrics, and baggage are three separate signals. Signals share a common subsystem – context propagation – but they function independently from each other.
```

Each signal consists of four types of packages:

> **API -** Public interfaces depended upon by 3rd-party libraries and application code instrumented with  OpenTelemetry. Every signal has its own, independent API package. These individual APIs are then bundled up into a Global API for convenience.

Reviewer 6: @arminru If I understood correctly, yes, this is a change from the current Specification (otherwise hard to support these experimental packages).

> **API -** Public interfaces depended upon by 3rd-party libraries and application code instrumented with  OpenTelemetry. Every signal has its own, independent API package. These individual APIs are then bundled up into a Global API for convenience.

Author: Yes, this is a change. 

It is still possible to offer a single unified API package, but experimental signals cannot be added to it until they are stable. In the meantime, experimental signal APIs must be required separately from this "global API."

So, in practice, the API is now broken up into separate packages. A unified API package with everything stable in it can can also be offered, as a convenience.

> 

Reviewer 6: Global API package?

Author: This needs some more detail, we don't have a term for this in the spec right now. But, in general, it looks like languages follow a pattern: implement each signal separately, in their own packages. Then, implementations combine expose access to all of the combined signals in another package. Starting to name that package the "Global API," as it is a reasonable design but has implications for versioning.

> **API -** Public interfaces depended upon by 3rd-party libraries and application code instrumented with  OpenTelemetry. Every signal has its own, independent API package. These individual APIs are then bundled up into a Global API for convenience.

Reviewer 4: ```suggestion
**API -** Public interfaces depended upon by 3rd-party libraries and application code instrumented with OpenTelemetry. Every signal has its own, independent API package. These individual APIs are then bundled up into a Global API for convenience.
```

> **SDK -** The implemention the API. The SDK is only referenced during setup and teardown of an application. SDKs include additional interfaces, such as plugin and lifecycle interfaces, which are not considered part of the API package as they are not imported by shared libraries and application code.

Reviewer: > The SDK is only referenced during setup and teardown of an application.

The SDK is active/involved in the entire runtime of the application. Do you mean referenced as in its methods are only called by the user during setup and teardown?

Author: Yeah I have to change the wording here. 

The intention is to say that SDK components are not cross-cutting concerns. They should not be directly referenced from within business logic, shared libraries, or instrumentation code. If the instrumentation for your MySQL client is interacting with z-pages or dynamically configuring your SDK there is a problem.

> **SDK -** As the implemention the API. The SDK is only referenced during setup and teardown of an application. SDKs include additional interfaces, such as plugin and lifecycle interfaces, which are not considered part of the API package as they are not imported by shared libraries and application code.

Reviewer 2: ```suggestion
**SDK -** As the implementation the API, the SDK is generally only referenced during setup and teardown of an application. SDKs include additional interfaces, such as plugin and lifecycle interfaces, which are not considered part of the API package as they are not imported by shared libraries and application code.
```

Reviewer 2: SDKs generally (I think?) provide dynamic tracing configuration (sampling configuration, for instance), so they will be referenced during the ongoing running of the application if that feature is used.

Author: To be clear, ongoing but in the background, correct? Not within app or library code as a dependecy?

Reviewer 2: I'm not sure. If we provide the SDK APIs to update tracing configuration, this could be done by any application in any way that they want to. Not by instrumentation, surely, but definitely by application code. The zpages extensions do, this, for example.

Author: Well, if we are creating SDK interfaces and are encouraging out end users to take dependencies on them in application code, AKA within transactions or mixed in with business logic, or encouraging shared libraries to make calls directly to the SDK, then that is a big problem. That code is no longer "SDK" code, that is part of our public interface, and must be put into the API. 

Is that the issue here, or are you saying there may be background processes within an application which run periodically, but are only managed at setup and do not get mixed into application code as a cross cutting concern?

Reviewer 2: I think what you're saying in the comment makes sense, but I don't think that means "only referenced during setup and teardown". What you're saying is that "application code involving transactions and business logic" shouldn't depend on the SDK, which I agree with. However, operator-oriented code will and should probably use the SDK to fulfill their more dynamic needs around how the SDK is functioning.

Author: Ok thanks, I'll clarify this.

> Each signal provides a mechanism for software to describe itself. A codebase, such as an API handler or a database client, takes a dependency on various signals in order to describe itself. OpenTelemetry instrumentation code is then mixed into the other code within that codebase. This makes OpenTelemetry a **cross-cutting concern** - a piece of software which must be mixed into many other pieces of software in order to provide value. Cross-cutting concerns, by their very nature, violate a core design principlie – seperation of concerns. As a result, OpenTelemetry requires extra care and attention to avoid creating issues for the codebases which depend upon it for instrumentaion.

Reviewer 5: > the codebases which depend upon it for instrumentaion

Nit: something a la `depend upon it for self-description` may be better. Applications don't use Otel for instrumentation. As you said at the beginning of this section `takes a dependency on various signals in order to **describe itself**`

Reviewer 2: Not sure what you mean by "Applications don't use Otel for instrumentation". I would say they do that a lot, when the application is adding manual instrumentation to its own code using otel APIs. do you not call that instrumentation?

Author: I agree with John, how is it that applications don't instrument themselves? 

Anyways, I did try to reword this to clarify that it is the cross-cutting nature of the APIs which make them so heavily depended upon.

"OpenTelemetry requires extra care and attention to avoid creating issues for the codebases which depend upon these cross-cutting APIs."

Let me know if that is better.

Reviewer 5: It's nitpicking the verbiage, feel free to ignore :) 

First sentence is "Each signal provides a mechanism for software **to describe itself**".
Last sentence is "codebases which depend upon it for **instrumentaion**"

But they both talk about one and the same action/process/whatever.

> **Semantic Conventions -** a schema defining the attributes which describe common concepts and operations which the signal observes. Note that unlike the API or SDK, Stable conventions for all signals may be placed in the same package, as they are often useful across different signals.

Author: yes, good catch

> OpenTelemetry is designed to seprate the portion of each signal which must be imported as cross-cutting concerns from the portions of OpenTelemetry which can be managed independently. OpenTelemetry is also designed to be a pluggable framework. To accomplish this these goals, each signal consists of four types of packages:

Reviewer 5: ```suggestion
OpenTelemetry is designed to separate the portion of each signal which must be imported as cross-cutting concerns from the portions of OpenTelemetry which can be managed independently. OpenTelemetry is also designed to be a pluggable framework. To accomplish this these goals, each signal consists of four types of packages:
```

> **Contrib –** plugins and instrumentation that make use of the API or SDK interfaces, but are not part of the core packages necessary for running OTel. The term contrib specifically refers to the plugins and instrumentation maintained by the OpenTelemetry SIG; it does not refer to third party plugins hosted elsewhere.

Reviewer: > The term contrib specifically refers to the plugins and instrumentation maintained by the OpenTelemetry SIG

Which SIG?

Reviewer 3: I'm assuming this references the language's SIG that maintains plugins and instrumentation for the particular language?  Agree this could use some clarity.

Reviewer 6: SIGs?

Author: Yes sorry, I want to say that "contrib" only refers to "the plugins each language SIG maintains as an official part of OpenTelemetry" and does not refer to "3rd party plugins maintained elsewhere."

> **Contrib –** plugins and instrumentation that make use of the API or SDK interfaces, but are not part of the core packages necessary for running OTel. The term contrib specifically refers to the plugins and instrumentation maintained by the OpenTelemetry SIG; it does not refer to third party plugins hosted elsewhere.

Reviewer 6: Instrumentation for libraries (Spring, Django, Express) are expected to rely on the API, but plugins/extensions can depend on the (reference) SDK implementation.

> **Contrib –** plugins and instrumentation that make use of the API or SDK interfaces, but are not part of the core packages necessary for running OTel. The term contrib specifically refers to the plugins and instrumentation maintained by the OpenTelemetry SIG; it does not refer to third party plugins hosted elsewhere.

Author: Correct, I'll word this better.

> ## Signal lifecycle

Reviewer 6: As mentioned, this is a very well known used term that we have used a lot before and we will be using in the future, so we should put it right in the Specification, once and for all (either at some root section or in the glossary).

> ## Signal lifecycle

Reviewer 6: Resolving this as https://github.com/open-telemetry/opentelemetry-specification/pull/1291 adds "signal" to the glossary.

> **API -** API packages consist of the public interfaces used for instrumentation. Any portion of OpenTelemetry which 3rd-party libraries and application code depend upon. To manage different levels of stability, every signal has its own, independent API package. These individual APIs may also be bundled up into a shared global API, for convenience.

Reviewer 5: Do we say anything about mixing API packages with different versions? E.g. Can I use tracing 1.2 and metrics 1.0 in my app/library?

Author: No, please. While individual libraries may require different versions of the API, we are trying to say that the version of OpenTelemetry which is actually loaded by the application must be a coherent tested release; no mixing and matching. *Ideally* that means the version numbers align, so there is no confusion. The two exception we have found is that experimental APIs may need to be versioned as v0 in Go, and the SDK and API may need to have separate version numbers. Regardless, when you pick the version of your SDK, that SDK has a specific version of the API which must be loaded.

Generally, if new APIs are backwards compatible, there is no reason to mix and match.

Reviewer 5: But some signal may be "1.2 stable" and some "1.2-experimental", right?

> **API -** API packages consist of the public interfaces used for instrumentation. Any portion of OpenTelemetry which 3rd-party libraries and application code depend upon. To manage different levels of stability, every signal has its own, independent API package. These individual APIs may also be bundled up into a shared global API, for convenience.

Reviewer 2: There can be a third package of common code that is used by both (this is how Java is being structured).

> **API -** API packages consist of the public interfaces used for instrumentation. Any portion of OpenTelemetry which 3rd-party libraries and application code depend upon. To manage different levels of stability, every signal has its own, independent API package. These individual APIs may also be bundled up into a shared global API, for convenience.

Reviewer 6: As John said, there can be a set of base packages. I imagine it will roughly look like this:

* Context propagation (no deps)
* Common (no deps)
* Baggage (depends on Common, Context)
* Tracing (depends on Common, Context)
* Metrics (depends on Common, Context)
... etc

> **API -** API packages consist of the public interfaces used for instrumentation. Any portion of OpenTelemetry which 3rd-party libraries and application code depend upon. To manage different levels of stability, every signal has its own, independent API package. These individual APIs may also be bundled up into a shared global API, for convenience.

Author: I agree, stable API packages can depend on other stable API packages. That is the joy of stability.

OpenTelemetry is structured around signals. Each signal represents a coherent, stand-alone set of functionality. Each signal follows a lifecycle.

![drawing](img/0143_api_lifecycle.png)

### Lifecycle stages

> **Experimental –** Bugs, breaking changes and performance issues may occur.

Reviewer: Bugs may occur at any stage 🙂 Maybe state bugs are more likely to occur here.

I'd also add that these components are usually not expected to be feature-complete at this stage.

Author: Good feedback, thank you :)

> **SDK -** The implemention of the API. The SDK is managed by the application owner. Note that the SDKs includes additional public interfaces which are not considered part of the API package, as they are not cross-cutting concerns. These public interfaces include constructors, configuration interfaces, and plugin interfaces.

Reviewer 5: Here we should better distinguish between application owner and library developer. The latter should not touch an SDK, it is only for the former.

Reviewer 2: Absolutely agree, @iNikem. In addition, we need to make sure that the SDK specification clearly lays out what the supported extension points/supported SDK "APIs" must be for every SDK implementation. 

There is a difference between extending an SDK, and configuring an SDK. These should be two classes of SDK APIs that we need to consider how we're going to manage from a versioning/release process.

Off the top of my head, the following is the list of what probably needs to be supported with strict ABI as potential extension points/apis in the SDK: Exporters, SpanProcessors, IdGenerators, Resource Semantic Attributes, Exportable SpanData, ReadableSpan, ReadWriteSpan, Sampler [I'm sure I'm missing a few].

Author: I totally agree with this sentiment, but the wording is a little difficult. Specifically, how do we differentiate between application business logic which should only touch the API, and the kind of SDK management subsystems which the application owner may write? Both types of code are technically "application code" written by the "application owner."

Reviewer 2: Advanced Samplers may be written and supported by someone other than the Application owner, and pulled in as a dependency at startup time, though.

**Stable –**  Stability guarantees apply, based on component tpye (API, SDK, Conventions, and Contrib). Long term dependencies may now be taken against these packages.

**Deprecated –** this signal has been replaced but is still supported under LTS.

**Removed -** a deprecated signal is no longer supported, and is removed.

All signal components may become stable together, or one by one in the following order: API, SDK, Semantic Conventions, Contrib.

### Stability

> Once a signal component is marked as stable, the following rules apply until the end of that signal’s existence.

Reviewer: Does this only apply to each signal as a whole or also feature by feature within a signal?

Author: It applies by signal. It's not possible to have experimental features embedded in a stable API due to the way it would break things like ABI guarantees. If it's experimental, it needs to be in a separate package.

It is feasible for a signal to become stable component by component: first the API, then the SDK, then conventions and then contrib. But I'm not going to focus on that in this proposal as I think it would be confusing.

**API Stability -**
No backward-incompatible changes to the API are allowed unless the major version number is incremented. All existing API calls must continue to compile and function against all future minor versions of the same major version. ABI compatibility is offered in languages which require it.

**SDK Stability -**

> Public portions of the SDK (constructors, configuration, und-user interfaces) must remain backwards compatible. Internal interfaces are allowed to break; ABI compatibility is not required.

Reviewer 6: > it seems like the SDK has an API by itself

That's correct (IIRC)

> Public portions of the SDK (constructors, configuration, und-user interfaces) must remain backwards compatible. Internal interfaces are allowed to break; ABI compatibility is not required.

Author: Yeah, we are hitting some limits here with the term "API." :P

There _are_ SDK interfaces which must remain stable. But, unlike with the instrumentation API, we don't require ABI compatibility and we don't need these SDK interfaces to be in separate packages.

I am not as concerned about SDK compatibility, but we should still avoid any breakage, especially in the portions of the SDK that end users interact with.

> Public portions of the SDK (constructors, configuration, und-user interfaces) must remain backwards compatible. Internal interfaces are allowed to break; ABI compatibility is not required.

Reviewer: ```suggestion
Public portions of the SDK (constructors, configuration, end-user interfaces) must remain backwards compatible. Internal interfaces are allowed to break; ABI compatibility is not required.
```

> Public portions of the SDK (constructors, configuration, end-user interfaces) must remain backwards compatible. Internal interfaces are allowed to break; ABI compatibility is not required.

Reviewer 4: ```suggestion
Public portions of the SDK (constructors, configuration, end-user interfaces) must remain backwards compatible. Internal interfaces are allowed to break, i.e., API or ABI compatibility is not required for internal interfaces.
```

This is the same as API right? I don't see any difference in the requirement that public API must be compatible, while internal APIs are allowed to break, with regards to API or SDK. If the intent is that the requirement is softer for SDKs, I don't think we can do that - custom samplers / exporters are good examples of artifacts that will exist outside of OpenTelemetry and that we must not break because of changes to ABI (following the principal of "Ensure that end users stay up to date with the latest release."). Even for just users of management interfaces, we can expect libraries like Spring - we should not create dependency conflicts between versions of Spring and the latest version of OpenTelemetry by having breaking changes.

So I'm not sure I see a reason to split API and SDK stability guarantees into two paragraphs like this, they're the same I think.

Some related conversation at https://github.com/open-telemetry/opentelemetry-java/pull/2212/files#r537932855

Author: We're skeptical that we can maintain SDK guarantees as easily as we maintain API guarantees. That's why they are separate.

Spring should never depend on the SDK. Why would that happen?

Reviewer 4: Spring _Sleuth_ (I should have clarified) is a configuration system for observability of Spring apps - users will set some flags in yaml or properties files and spring will configure the SDK on behalf of the user based on those files or what's in the classpath. It has a very similar role to the java autoinstrumentation agent, and does need to depend on the SDK. Unlike the autoinstrumentation agent, it doesn't have the power to bridge the app into an isolated SDK and will depend on our versions.

That being said, it could be possible to allow Sleuth and SDK versions to be tied since they're at the top level - though this needs to be confirmed.

`Sampler` and `SpanExporter` are more important - there definitely will be custom implementations of these. At the Java SIG meeting, we had the thought that we need to identify the SDK APIs that are extension points and we do need to make sure don't cause version conflicts.

Reviewer 7: @anuraaga In the case of Java, I had created https://github.com/open-telemetry/opentelemetry-java/issues/1909 which I think would help with the breakage of anything outside the SDK

Author: Totally agree that the SDK needs some stability. We absolutely need to mark out what counts as public and backwards compatible in the SDK.

Reviewer 8: I don't believe that we should enforce complete ABI compatibility for "all SDK" implementations (ie. languages), so I believe that the original comment of "ABI compatibility is not required." is the correct way to phrase this here and any specific language / SDK implementation can then "choose" to add further constraints/support on their provided implementations.

But in general, we should only be guaranteeing that the "public" API's (entry points, configs) are stable as already pointed out in the preceding points with "internal" (implementation level) details (API, Class, functions etc) should always be up for grabs otherwise (depending on the language) you will end up in code bloat city with super fragile implementations.

Author: I clarified that ABI compatibility requirements are left to individual languages.

> **Stable –**  Stability guarantees apply, based on component tpye (API, SDK, Conventions, and Contrib). Long term dependencies may now be taken against these packages.

Reviewer 5: ```suggestion
**Stable –**  Stability guarantees apply, based on component type (API, SDK, Conventions, and Contrib). Long term dependencies may now be taken against these packages.
```

what does "based on component type" mean? Do stability guarantees differ somehow between types?

Author: @iNikem yes stability guarantees may vary between component types. Contrib, API, and SDK have details around stability which are specific to those types of components.

**Semantic Conventions Stability -**

> Semantic Conventions may not be removed once they are stable. New conventions may be added to replace usage of older conventions, but the older conventions are never removed.

Reviewer 6: Probably we should deprecate them only. IIRC there will be an actual 'feature removal' policy in the future, so we can consider our options here.

> Semantic Conventions may not be removed once they are stable. New conventions may be added to replace usage of older conventions, but the older conventions are never removed.

Reviewer: ```suggestion
Semantic Conventions may not be removed once they are stable. New conventions may be added to replace usage of older conventions, but the older conventions are never removed, they will only be marked as deprecated in favor of the newer ones.
```

> Semantic Conventions may not be removed once they are stable. New conventions may be added to replace usage of older conventions, but the older conventions are never removed.

Author: Yes, there is truly no reason to remove old constants, or to forget that there is code out there which may report using old values.

> **Deprecated –** this signal has been replaced but is still supported under LTS.

Reviewer 5: You have never mentioned "LTS" in this doc before. What is it?

Reviewer 9: Probably "Long term Support".

Reviewer 5: It is, but this still needs to be defined before first usage.

Reviewer 10: +1.  "support" is also mentioned below in the deprecation section.

Author: Derp, I was actually trying to remove any reference to LTS, looks like I missed one.

**Contrib Stability -**

> Plugins and instrumentation are kept up to date, and are released simultaneously (or shortly after) the latest release of the API. The goal is to ensure users can update to the latest version of opentelemetry, and not be held back by the plugins that they depend on.

Reviewer 3: That would make sense to me, but it might cause a build headache.  Not sure what the long term intent is here.

> Plugins and instrumentation are kept up to date, and are released simultaneously (or shortly after) the latest release of the API. The goal is to ensure users can update to the latest version of opentelemetry, and not be held back by the plugins that they depend on.

Author: Ah, good clarification. I did not intend to say that contrib packages must have matching version numbers. Contrib packages have their own version numbers and are released individually as needed. 

My intent was to say that a release of OpenTelemetry is not complete until contrib packages are also updated as needed, to reflect the latest conventions and to make use of the latest API.

> Plugins and instrumentation are kept up to date, and are released simultaneously (or shortly after) the latest release of the API. The goal is to ensure users can update to the latest version of opentelemetry, and not be held back by the plugins that they depend on.

Reviewer 5: > My intent was to say that a release of OpenTelemetry is not complete until contrib packages are also updated as needed

Does this assume some kind of central coordination of a "release of OpenTelemetry" across language implentations?

Public portions of contrib packages (constructors, configuration, interfaces) must remain backwards compatible. ABI compatibility not required.

> Telemetry produced by contrib instrumentation must also remain stable and backwards compatible, to avoid breaking alerts and dashboard.

Reviewer 2: I'm not sure what this ends up implying. What does it mean in practice? If a bug is found in some instrumentation, and the results are different telemetry, does that instrumentation have to have a major version bump?

Reviewer 6: @jkwatson Interesting point. I think bug fixing should be the only valid reason to 'break' this invariant.

Author: @jkwatson that sounds reasonable to me. Contrib packages should have their own version numbers, and if they break existing data, then it's a major version bump for that package. But I'm not sure how painful this is to manage in practice. 

Maybe there is some practical advice to be learned from New Relic, Dynatrace, etc, about managing the balance the between preserving existing data and making changes and improvements to instrumentation.

Reviewer 2: I'm just thinking of the opentelemetry-java-instrumentation agent. If every change in telemetry led to a new major version of the agent, I think that might get quite unwieldy.  @trask opinions?

Reviewer 11: @jkwatson ya it's tricky, let's discuss with @iNikem and @anuraaga in [tomorrow night's SIG meeting](https://calendar.google.com/event?action=TEMPLATE&tmeid=Nm1vZXBtNmxxMm44dW9sOWRmZWwwdm5icnRfMjAyMDEyMDlUMDYwMDAwWiBnb29nbGUuY29tX2I3OWUzZTkwajdiYnNhMm4ycDVhbjVsZjYwQGc&tmsrc=google.com_b79e3e90j7bbsa2n2p5an5lf60%40group.calendar.google.com)

@tedsuo if it's not too late, would be great to have you join

I'll post back here on Wed with thoughts from the team

Reviewer 12: Do we have a good handle on what "compatible telemetry" means?   I believe the definition is actually vendor / backend-specific.   

E.g. "Is it acceptable to add a new label to a metric produced by a contrib module"?  The strictest definition of "compatible" would be 'never', but *likely* we can be more lenient here and provide a path for this telemetry to slowly evolve to account for new use cases or issues with previous versions.

Reviewer 5: > The strictest definition of "compatible" would be 'never'

That's not actually true. "Compatible" is defined by Semver.

Reviewer 2: But if breaking changes depend on the backend, the instrumentation authors may not even know what is breaking for one vendor/backend vs. another.

Author: Dang, sorry @trask , I was on the road and just missed the meeting.

I was not considering that the agent bundles up instrumentation in Java. I think whatever you decide as a group will be more informed than my pot shots, but it does seem more useful for the version of the agent to be separate from the version of the plugins. But I can see how decoupling the two could really increase maintenance costs. 

If users have to update to the latest plugin to get the latest agent, then breaking instrumentation data should probably bump the major version number. But again, I think the Java crew can thread the needle on this better than I can.

@jsuereth I suggest that versioning for instrumentation data should just be like semver and the rest of OTel. Adding more data is backwards compatible, removal and mutation of existing data is not. 

@jkwatson In the case of a bug, fixing it would be a patch, unless the "bug" we're talking about is some kind of useful data which users rely upon. But that sounds like something other than a bug. I think some subjective reasoning would be needed in each case, to decide on whether anyone would realistically be dependent on the bug. The other option is to be pedantic, but I'm not sure if that would better serve anyone's needs in this case.

Reviewer 12: @tedsuo Regarding instrumentation, I guess i'm trying to understand better.   Here's some scenarios:

1. I add a new label to a metric in a importer/collector, by this definition it's not "incompatible"
2. On some backends, this new label may break existing dashboards, queries, alerts or other usage of that metric.  This would be unexpected if I thought the version bump was compatible.

What I'm asking for is a *clear* definition of what "backwards compatible" means for metrics.   Should all interaction with that metric in a downstream system be backwards compatible?  I think we'll need to define what compatible metric changes are (in OpenTelemetry's definition) so vendors can adopt their own adaptations to this if their backends don't have similar compatibility.

I may split this discussion into a separate thread with some more thoughts.

Reviewer 11: the plan from java-instrumentation is to keep all artifacts (plugins, javaagent) at the same version, massively deprecate if we have to, but not introduce any breaking changes until OTel Java API goes to v2, at which point the java-instrumentation repo would go to v2 as well.

for experimental (non-GA) instrumentation artifacts, the preference is to use version suffix, e.g. `1.15.0-alpha` for any experimental artifacts released as part of `1.15.0` release train.

Author: @jsuereth I don't see how adding a new label would break a dashboard? 

If you have a more nuanced proposal I'm happy to discuss it. But for now, I am going to stick to "removal and mutation of existing data" as the clear definition of backwards compatibility. Users can expect that additional spans, metrics, attributes, and events may be added and take that into account when upgrading.

Reviewer 12: @tedsuo This may be a GCP specific thing, but adding a label where one did not exist previously can alter how timeseries are returned, depending on what aggregation you're performing in the query.   

IF the label-set is different and you write a query/alert per-label-set, you can window up with a scenario where your timeseries forks from being *one* set of points to N (based on the cardinality of the new label), which could lead to alerts.  This could affect the threshold of an alert operating on raw timeseries, or not aggregated across that label.

From an OpenTelemetry perspective, no data is lost.  From a user perspective, they just upgraded a library and suddenly  they have to adjust their queries to make sure they're aggregating this new label and not operating on raw timeseries.

Author: ^ cc @jmacd

> Public portions of the SDK (constructors, configuration, end-user interfaces) must remain backwards compatible. Internal interfaces are allowed to break; ABI compatibility is not required.

Reviewer 10: If the API can make breaking changes when it increases its major version, presumably the SDK must also have a breaking change, which isn't mentioned here.

I think you could simplify this section by talking about "Code Stability" vs "Semantic Conventions Stability".  The same rules for backwards-compatibility should apply to all kinds of code.

Author: 🤷‍♀️   People keep asking what guarantees specifically apply to semantic conventions.

### Deprecation

In theory, signals could be replaced. When this happens, they are marked as deprecated.

Code is only marked as deprecated when the replacement becomes stable. Deprecated code still abides by the same support guarantees as stable code. API’s remain stable and backwards compatible.

### Removal

> Packages are end-of-life’d by being removed, along with a major version bump.

Author: Does this make sense:
The release gets a major version bump, and no longer contains the package. The last version of the deprecated  package is the one that was in the final release, prior to the major version bump. No more versions of the removed package are ever made, except when there are security and bug patches backported to the prior major version.

> We currently have no plans for deprecating signals or creating a major version past v1.0.

Reviewer 3: What sort of criteria would require a major version past v1.0?  I think in a previous meeting we discussed an additional signal beyond tracing, metrics, and logging?

Author: Right now, there are no plans for moving past v1. 

I backed away from additive changes like new signals or stable releases creating a major version bump, as that broke too many assumptions. Only the removal of existing code which requires backwards compatibility would create a major version bump. 

In theory, we could develop a "v2.0" of some kind which deprecated enough existing code that we would want to eventually remove the old code.

Reviewer 5: Wow, are we really thinking that we are creating the perfect OpenTelemetry with v1.0? And we don't plan for v2.0? I understand that we want 1.0 to be solid and long-lived, but saying that it will be the only and final version is... too optimistic?

Reviewer 2: I don't think this is saying there won't be 2.x versions. It just says that we don't have any plans to release breaking changes at this point. We very well may have them later, for sure. But, not *currently*.

Author: What @jkwatson said. @iNikem we have a situation where our users are going to rely on v1.0 for years. So, while we can add new things – including a better, totally different tracing API – we can't remove the old things or cease to support them. So that "totally different tracing" API would have to be an entirely new signal which coexists and interoperates with the current tracing signal. v2.0 would mark the end of support for current tracing, not the addition of new awesome tracing. And we don't want to ever end that support, if we could help it.

And to be clear if you think this is crazy – consider that we are already doing exactly this with the OpenTracing API. We invented a new tracing API but continue to support the old one.

Reviewer 5: Well, if you put it that way, then it makes more sense :)

Reviewer 8: So what about v 1.x, are we saying that there will never be a next "part" release?
Arguing that we are only ever going to create a ""totally different tracing" (which wraps the previous one) I believe is missing the reality that we WILL find incremental changes that can be safely added without breaking backward compatibility...

Author: Yes of course. Minor version bumps are fine, provided they do not break backwards compatibility.

Reviewer 13: Can we be a bit more specific with minor version bumps. For example:
- **v1.0** - apps compiled against **v1.0** are also API- and ABI-compatible with anything **v1.x**
- **v1.1** - may contain additive changes, e.g. new API (_as in new API in a separate new class for C++_) , considering that any new additions are not breaking ABI compatibility.

Re. **Logging API** specifically - assuming that the Logging is NOT part of GA for **v1.0** , is Logging gonna be part of **v2.0** or **v1.1** ? With a provision of additive changes being OK, assuming these are not ABI-breaking, we'd be able to release Logging in **v1.1** rather than bumping up the major version.

What is your opinion on this?

Author: Correct, adding new signals is a minor version bump. So metrics is a minor bump, and logging is a minor bump. All additive changes are a minor version bump.

## Version Numbers

OpenTelemetry follows [semver 2.0](https://semver.org/) conventions, with the following distinction.

> For the purposes of versioning, all code within a release is treated as if it were part of a single package. All packages within a release receive the same version number. By keeping version numbers in lockstep, end users can understand which packages belong to which release.

Reviewer 14: I understand this as every package getting a new version number when one package gets an update regardless if every package has had any changes or not.

This could introduce some issues. Semver 2.0 specifically says that increments are to be done to version numbers when incompatible API changes are made, functionality is introduced in a backwards compatible manner or bugs are fixed in a backwards compatibility manner. If a package has its version number increased without this happening, then the meaning of the version number is lost, since the user can't now tell if a package that went from X.Y.Z to X.Y.Z + 1 actually has a bug fix or not.

This also may cause problems when debugging issues. If version numbers are increased only when packages have changed then the user who is facing an issue can downgrade a package that may be causing this issue. If the issue persists, then the user knows that the changes in that package are not causing the issue. If version numbers are increased even if packages have no changes, now the user can't use version numbers for debugging purposes, because if the user downgrades a package and the issue persists, then the user can't tell if the issue is related or not because it is not possible to tell from the version number if the package has changed or not.

Reviewer 6: > Semver 2.0 specifically says that increments are to be done to version numbers when incompatible API changes are made

That happens only when the major version changes. As long as we stick to minor/patch increases, we should be all fine.

> an user won't be able to tell if a package that went from X.Y.Z to X.Y.Z + 1 actually has a bug fix or not.

I'd argue CHANGELOG should be the reference for users in this case ;)

Author: Yes, we are saying, specifically, that semver applies to all the code in core, at once. Individual sub-packages are not separate code bases with separate version numbers. There is only one codebase, **opentelemetry-core**, which contains all the code for that language implementation. So users should look in the changelog when they update to the latest version.

Contrib packages, on the other hand, are stand-alone and can be versioned independently.

Reviewer 14: @carlosalberto Actually Semver 2.0 specifically says that _increments are to be done to version numbers when incompatible API changes are made, functionality is introduced in a backwards compatible manner or bugs are fixed in a backwards compatibility manner_. What I mean with this is that increments are to be done to version numbers (major, minor and patch numbers) when code changes happen, so even if we only increase non-major version numbers it would be Semver 2.0 non-compliant to increase **any** number (major, minor or patch) if there is no change in the code associated with that version number increase.

Reviewer 14: @tedsuo, if I understand correctly, the intention of keeping the version number in lockstep for all _core_ OpenTelemetry components, is to make it possible for users to understand which packages belong to which release. Now, maybe we are trying to tie the _OpenTelemetry release_ concept to the _OpenTelemetry individual package version number_ concept when they should be handled separately? As far as I understand, the latter concept has a meaning clearly defined already in Semver 2.0. The former concept may mean "a collection of specifically versioned individual OpenTelemetry individual packages"? :shrug:

I mean, maybe what we are trying to achieve here is to have another number, an _OpenTelemetry release_ number? For example, the OpenTelemetry Java release 7 means the collection of the OpenTelemetry Java API 3.5.1, OpenTelemetry Java SDK 4.5.5, and OpenTelemetry Java SomethingElse 8.1.3 packages?

We can give some meaning to the release number by saying _**this** set of tests was run using **this** release_ (which means, with these specific individual packages) so that we are 110% sure that everything works and is absolutely perfect and completely and absolutely bug-free and the bestest software ever (:stuck_out_tongue_closed_eyes: )?

Reviewer 14: Also, I have some doubts regarding expecting users to rely on the CHANGELOG to find out if there were actually changes made in the individual packages. I would be a bit annoyed if I had to read CHANGELOG after CHANGELOG file to find out which was the last package that included a change when I am debugging something in the scenario described in my first comment and I would be asking myself if some tool that relies on version numbers to automate some task will work with OpenTelemetry.

Reviewer 5: @ocelotl Do you propose "Release Trains"?

Reviewer 6: > This could introduce some issues. Semver 2.0 specifically says that increments are to be done to version numbers when incompatible API changes are made, functionality is introduced in a backwards compatible manner or bugs are fixed in a backwards compatibility manner.

I feel we are running in circles. Simply: for 1.x versions, changes are expected to be have additive changes, and we won't introduce breaking changes.

> Also, I have some doubts regarding expecting users to rely on the CHANGELOG to find out if there were actually changes made in the individual packages. I would be a bit annoyed if I had to read CHANGELOG after CHANGELOG 

That's not different that when you upgrade to a new Java/DotNet versions, as you need to read the documentation (CHANGELOG may not be the most friendly one, but still). 

> to find out which was the last package that included a change when I am debugging something in the scenario described in my first comment 

You are trying to detect a bug based on the fact changes happened in a dependency? What if 15 packages included changes, but they are unrelated or code bases clean up? Do you still check them? If this is not the case, please elaborate.

Reviewer 14: @iNikem Not necessarily :sweat_smile: I am simply suggesting that we could group individual packages and assign a number to this group so that the versioning numbers of individual packages can keep their usual meaning. Release Trains includes a lot more than just this, like scheduling, commitments, specific roles for participants, etc.

Reviewer 14: @carlosalberto hmm, maybe I am not being clear, sorry. I just mean that it is not Semver 2.0 compliant to increase a version number if there are no changes associated with that increase. I think the root of this issue/misunderstanding is the concept of OpenTelemetry release. @tedsuo, what do you mean with _release_ in this statement:

> By keeping version numbers in lockstep, end users can understand which packages belong to which release.

?

Author: A release is a tested version of opentelemetry core. It is a tagged version of the core repo, and the resulting artifacts generated by that CI pipeline. Contrib packages are separate, but required to be compatible with the latest release.

> By keeping version numbers in lockstep, end users can understand which packages belong to which release.

By this, I mean that if users are bumping versions of individual subpackages, rather than bumping the entire core, then they can end up running an untested combination of packages. If we version every subpackage separately, then it is much harder for users to understand when they are running an official release, and when they are using an untested combination. OpenTelemetry is easier to comprehend when it is treated as a monolithic codebase.

But, based on many of these conversations, I'm coming around to the idea that the API and SDK may benefit from versioning separately, and the SDK marks breaking changes with major version bumps. This separation will make life difficult for maintainers. In most languages, it will probably result in splitting core into two repos. It has been simpler for *everyone* to treat OpenTelemetry as a single monolithic codebase. That said, I can see how in the long term it would be helpful to have an API version number and an SDK version number. This would allow the SDK to follow semver, and allow  breaking changes without implying the instrumentation API has changed.

But every subpackage of the SDK definitely should have the same version number. The SDK should be treated as a monolithic codebase in this regard. Experimental packages may be isolated and kept at v0 until they are stable. But  stable packages should be versioned together.

Note: there may be packages in core which should really be in contrib, such as exporters. For example, It may be reasonable for the prometheus exporter to keep its own version number, in which case it should be moved out of the SDK. But if it needs to be in the core SDK, it's part of the monolith.

Reviewer 8: On this "helpful to have an API version number and an SDK version number." I think it's not just _"helpful"_, but **required**, if we assume the API surface is (long term) stable it doesn't make sense to rev the API version when there are no changes, it will just complicate the product development and potentially cause some products to just not upgrade because of potential breaking changes (assuming they don't read the Changelog and it's not explicitly called out that nothing actually changed).

So as you point out "This would allow the SDK to follow semver, and allow breaking changes without implying the instrumentation API has changed." I think this should be explicitly called out.

Reviewer 15: I'd like to strongly advocate that we allow the SDK and API version numbers to differ. Once stable, the API should have few if any changes. SDKs are likely to undergo a number of bugfixes, additions, improvements, etc. Revving the API version to match the SDK version when there have not been any changes to the API is a very confusing signal. We hope and expect to have many third parties depending on the OpenTelemetry API over time. This potentially includes framework authors who are building in OpenTelemetry instrumentation hooks and alternative SDK implementors. Coupling the API version with the OpenTelemetry SDK could be challenging for third parties that depend on only the API.

Reviewer 16: Strong :+1: to @mwear suggestion to keep the API version separate. We discussed this at length in the JS SIG today and I fully agree that keeping the API as unchanging as possible makes it significantly easier for package maintainers to build in support for opentelemetry, which is the goal.

Reviewer 17: I am very strongly in favor of separate versions for SDK and API. I am fairly certain we had this discussion & conclusion almost a year ago, and are just rehashing those arguments.

Author: Great. Glad we're coming to consensus.

> ### Deprecation

Reviewer 10: There isn't any time-period described here between deprecation and removal, which makes deprecation somewhat moot.  I can simply replace old with new, and cut a new major version without deprecating.  If this is binding in some form (e.g. through a minimum "deprecation period"), we should spell that out.  If not, make it clearer that this is an optional, be-a-nice-community sort of thing.

Author: So, I don't want to arbitrarily pick a minimum deprecation period, and no one has pointed to any kind of convention or standard to follow. The current deprecation period should be considered "infinite" for the time being.

Reviewer 10: I'm not quite following.  IIUC, an infinite deprecation period would imply we can't remove signals, even with a new major version, which doesn't align with the rest of the proposal.

Here is what I would suggest as a definition for deprecation:

"A deprecated signal is a signal which will be removed in the next major version, but is identical to a stable signal in all other respects.  Marking a signal deprecated does not imply a new major version must be cut, but **guarantees** that it will be removed if/when a new major version is cut.  All signals which are removed in a new major version must be marked deprecated in the old major version before releasing the new major version--no cherrypicking deprecations after-the-fact, and no removing signals which are not deprecated. Marking a signal deprecated is intended to smooth the transition to a new major version by encouraging users to move off of the deprecated signal prior to upgrading to the next major version.  A user who has moved off of all deprecated signals will not encounter any compilation errors when upgrading to the new major version."

Note that contrib packages are released separately from core packages. Contrib releases are not required to match version numbers with the core releases.

> In some languages, package managers react poorly to experimental packages having a verion higher than 0.X. In these cases, a language-specific workaround is required.

Reviewer: ```suggestion
In some languages, package managers react poorly to experimental packages having a version higher than 0.X. In these cases, a language-specific workaround is required.
```

> Code is only marked as deprecated when the replacement becomes stable. Deprecated code still abides by the same support guarantees as stable code. Deprecated APIs remain stable and backwards compatible.

Reviewer 6: I think the idea is to actually remove those components _way down the road_, just not right away (we did this in OpenTracing, removing APIs after deprecating them, and there were a few complains).

> Code is only marked as deprecated when the replacement becomes stable. Deprecated code still abides by the same support guarantees as stable code. Deprecated APIs remain stable and backwards compatible.

Author: Yeah, I can't see a reason for us removing stable features just because we decide we don't like them any more. In those cases, we just leave them alone and let them be.

> Code is only marked as deprecated when the replacement becomes stable. Deprecated code still abides by the same support guarantees as stable code. Deprecated APIs remain stable and backwards compatible.

Reviewer 10: Agreed.  The only process by which we should remove signals is by introducing a new major version.  We should also put a high price on new major versions, which pushes removal of signals way down the road.

I just don't want to imply that major version releases, which can remove signals, will have feature-parity with previous major versions.
```suggestion
If the signal is being replaced, code should only be marked as deprecated after the replacement becomes stable. Deprecated code still abides by the same support guarantees as stable code. Deprecated APIs remain stable and backwards compatible.
```

> Note: different languages implementations do not need to have matching version numbers, nor do implementations have to match the version of the specification they implement. For example, it is fine to have Python at 1.2.8, Java at 1.3.2, and the spec at 1.1.1. Within python, all packages in the 1.2.8 release have the version number 1.2.8, even if they are experimental.

Reviewer: ```suggestion
Note: different language implementations do not need to have matching version numbers, nor do implementations have to match the version of the specification they implement. For example, it is fine to have opentelemetry-python at 1.2.8, opentelemetry-java at 1.3.2, and the spec at 1.1.1. Within opentelemetry-python, all packages in the 1.2.8 release have the version number 1.2.8, even if they are experimental.
```
to make clear we are not referring to versions of the languages themselves but our OTel implementations

> Note: different languages implementations do not need to have matching version numbers, nor do implementations have to match the version of the specification they implement. For example, it is fine to have Python at 1.2.8, Java at 1.3.2, and the spec at 1.1.1. Within python, all packages in the 1.2.8 release have the version number 1.2.8, even if they are experimental.

Reviewer 18: >Within python, all packages in the 1.2.8 release have the version number 1.2.8, even if they are experimental.

I think one of the concerns @bogdandrutu had with this was that people will use an API without reading the notice that it is unstable.

If we are doing this we should spend an extra effort to make sure there is a good visibility of unstable APIs, that they are clearly tagged and labeled in the docs.

Author: Those users won't look at the version number either. :P 

Note: it looks like we can't do this in Go; the package manager really prefers that experimental packages be v0.

**Major version bump**  
Major version bumps only occur when there is a breaking change to a stable interface, or the removal of deprecated signals.

> OpenTelemetry values long term support. The expectation is that we will version to v1.0 once the first set of packages are declared stable. OpenTelemetry will then remain at v1.0 for years. THere are no plans for a v2.0 of OpenTelemetry at this time. Additional stable packages, such as metrics and logs, will be added as minor version bumps.

Reviewer: ```suggestion
OpenTelemetry values long term support. The expectation is that we will version to v1.0 once the first set of packages are declared stable. OpenTelemetry will then remain at v1.0 for years. There are no plans for a v2.0 of OpenTelemetry at this time. Additional stable packages, such as metrics and logs, will be added as minor version bumps.
```

**Minor version bump**  
Most changes to OpenTelemetry result in a minor version bump.

* New backward-compatible functionality added to any component.
* Breaking changes to internal SDK components.
* Breaking changes to experimental signals.
* New experimental packages are added.
* Experimental packages become stable.

**Patch version bump**  

> Patch versions make no changes which would require recompilation or potentially break a user. The following are examples of patch fixed. 

Reviewer 14: ```suggestion
Patch versions make no changes which would require recompilation or potentially break application code. The following are examples of patch fixes. 
```

> Note: different languages implementations do not need to have matching version numbers, nor do implementations have to match the version of the specification they implement. For example, it is fine to have Python at 1.2.8, Java at 1.3.2, and the spec at 1.1.1. Within python, all packages in the 1.2.8 release have the version number 1.2.8, even if they are experimental.

Reviewer 5: Do we want to say anything about the compatibility of the produced telemetry data between different language implementation and their versions? E.g if I have a system running in production where different parts use different languages and thus different Otel language implementations. Should I somehow coordinate versions of different language implementations or am I free to mix them at whim?

> * Breaking changes to internal SDK components.

Reviewer 10: This implies bug-fixes don't ever require backwards-incompatible changes to internal SDK components, which I don't think is the case.  For example, a fix may require passing an additional argument to a function.

Author: thanks, clarified.

* Bugs
* Security fixes
* Documentation

Currently, OpenTelemetry does have plans to backport bug and security fixes to prior minor versions. Security and bug fixes are only applied to the latest minor version. We are committed to making it feasible for end users to stay up to date with the latest version of OpenTelemetry.

## Open questions

### “OpenTelemetry GA” aka v1.0

In theory, we have assumed that tracing and metrics would be released together as a v1.0, which we refer to as “OpenTelemetry GA.”

However, in practice, it appears that tracing will be ready to GA before metrics. Tracing is ready today in .NET, and metrics are still months away from being finished.

> While we can continue to use the term OpenTelemetry GA to mean the release of both tracing and metrics, we should decouple this from our versioning and support terminology.  That allows us to announce stable tracing this month.

Reviewer 14: ```suggestion
While we can continue to use the term OpenTelemetry GA to mean the release of both tracing and metrics, we should decouple this from our versioning and support terminology. That allows us to announce stable tracing on December 2020.
```

> * Bugs

Reviewer 5: ```suggestion
* Bugs fixes which don't require minor version bump per rules above
```

> Currently, OpenTelemetry does have plans to backport bug and security fixes to prior minor versions. Security and bug fixes are only applied to the latest minor version. We are committed to making it feasible for end users to stay up to date with the latest version of OpenTelemetry.

Reviewer 10: In the hypothetical world in which we cut a new major version, and there was a security vulnerability, I think we would want to patch the latest minor version **in both major versions**.  It would be hard to force customers to upgrade to a new major version to get a security fix, as major version upgrades will be disruptive by definition.

Author: So, I agree that back-porting across major versions seems both reasonable and likely. But how many major versions? I would rather not pick something arbitrary, and leave it open for the future when we come to it. We can always add this clarification later.

Reviewer 10: In my doc I had proposed all major versions that have had a minor release in the past year, but starts to get into the LTS discussion, which may be too much for this proposal.  It might be easier to start with something optional, like "it is up to maintainers to decide whether to support additional minor versions, or to support minor versions from previous major versions"

> While we can continue to use the term OpenTelemetry GA to mean the release of both tracing and metrics, we should decouple this from our versioning and support terminology.  That allows us to announce stable tracing this month.

Reviewer 12: I like this goal, but I'd love to be more aggressive with it.  We should release components we know to be stable ASAP, and evolve in a compatible way as we go.

As you know, I'd love for us to launch the portion of metrics we consider stable as well, but it may be more work to clearly denote them than it is to just finish up the remaining work on metrics.

Reviewer 2: Without a complete metrics SDK spec, I don't think we can, in good faith, release a metrics implementation. Do you see some way we could?

Author: I agree, it is not realistic for a signal to be declared stable until its spec is stable. If the spec can be carved up in such a way that we can declare part of it stable, great.

> # OpenTelemetry versioning and stability

Author: Sure thing; it's true that we tend to be a little sloppy about what we are pointing to when we say "OpenTelemetry."

OpenTelemetry is a large project with strict compatibility requirements. This proposal defines the stability guarantees offered by OpenTelemetry, along with a versioning and lifecycle proposal which defines how we meet those requirements.

> Language implementations are expected to follow this proposal exactly, unless a langague or package manager convention interferes significantly. Implementations must take this cross-language proposal, and produce a language-specific proposal which details how these requirements will be met.

Reviewer 19: OpenTelemetry does use RFC 2119, but this is an OTEP which can have a more vague wording. The expectation is that once the OTEP is approved, a new document with more precise wording where necessary is added to the spec based on it.

> Language implementations are expected to follow this proposal exactly, unless a langague or package manager convention interferes significantly. Implementations must take this cross-language proposal, and produce a language-specific proposal which details how these requirements will be met.

Reviewer 8: nit typo: langague

Author: That's it! I'm off to find a spellchecker for vscode.

> Language implementations are expected to follow this proposal exactly, unless a langague or package manager convention interferes significantly. Implementations must take this cross-language proposal, and produce a language-specific proposal which details how these requirements will be met.

Author: Yes, correct. Once we have feedback from the language specific proposals currently being drafted, it will be clearer what can feasibly be listed as a hard requirement when we add this to the spec. 

But, I should emphasize that the ultimate goal of this OTEP is not strict compliance with a specific implementation approach. It is far more important that implementors understand the stability requirements, and meet those requirements. Achieving backwards compatibility is an area where programming languages differ greatly, in terms of language requirements and mechanisms. Some languages have ABI requirements; other languages don't even have types. So I expect some amount of divergence in the implementation details, but no divergence in the strict stability and compatibility guarantees we want to offer. 

If the stability and depreciation goals are not clear, please let me know so we can clarify further.

> Language implementations are expected to follow this proposal exactly, unless a langague or package manager convention interferes significantly. Implementations must take this cross-language proposal, and produce a language-specific proposal which details how these requirements will be met.

Reviewer 20: > OpenTelemetry does use RFC 2119, but this is an OTEP (proposal) which can have a more vague wording. The expectation is that once the OTEP is approved, a new document with more precise wording where necessary is added to the spec based on it.

@Oberon00 Hm... probably you meant "OpenTelemetry spec", not "OpenTelemetry", right? (I'm still struggling to figure out terms in this project :wink: )

Since this OTEP addresses recommendation/expectation/requirement, and the level of requirement is the core of this OTEP... I believe it should be more clear than now - although there are much more room in other types of OTEP. I don't think it's a good idea to delay "precise term" discussion to PR in the spec.

> I expect some amount of divergence in the implementation details, but no divergence in the strict stability and compatibility guarantees we want to offer.

@tedsuo I like the distinction indeed! However what are "strict" or "hard" requirement is not clear unless we use clear terms to describe them. That's why we have to use RFC 2119 even in this OTEP to make it clear.

> Language implementations are expected to follow this proposal exactly, unless a langague or package manager convention interferes significantly. Implementations must take this cross-language proposal, and produce a language-specific proposal which details how these requirements will be met.

Reviewer 6: > Hm... probably you meant "OpenTelemetry spec"

OTEPs are expected to be very general informal proposals (after one OTEP is accepted _then_ it has to be incorporated into the actual Specification, which is when it has to have the proper terms, etc). I suggest don't getting lost into the details now (we will have time for that after this part hits the Spec).

(Theoretically, APIs can be deprecated and eventually removed, but this is a process measured in years and we have no plans to do so.)

**Allow for multiple levels of package stability within the same release.**
Provide maintainers a clear process for developing new, experimental APIs alongside stable APIs. DIfferent packages within the same release may have different levels of stability. This means that an implementation wishing to release stable tracing today must ensure that experimental metrics are factored out in such a way that breaking changes to metrics API do not destabilize the trace API packages.

## Relevant architecture
![drawing](img/0143_cross_cutting.png)

At the highest architectural level, OpenTelemetry is organized into signals. Each signal provides a specialized form of observability. For example, tracing, metrics, and baggage are three separate signals. Signals share a common subsystem – context propagation – but they function independently from each other.

> Each signal provides a mechanism for software to describe itself. A codebase, such as an API handler or a database client, takes a dependency on various signals in order to describe itself. OpenTelemetry instrumentation code is then mixed into the other code within that codebase. This makes OpenTelemetry a **cross-cutting concern** - a piece of software which must be mixed into many other pieces of software in order to provide value. Cross-cutting concerns, by their very nature, violate a core design principlie – seperation of concerns. As a result, OpenTelemetry requires extra care and attention to avoid creating issues for the codebases which depend upon these cross-cutting APIs.

Reviewer 2: ```suggestion
Each signal provides a mechanism for software to describe itself. A codebase, such as an API handler or a database client, takes a dependency on various signals in order to describe itself. OpenTelemetry instrumentation code is then mixed into the other code within that codebase. This makes OpenTelemetry a **cross-cutting concern** - a piece of software which must be mixed into many other pieces of software in order to provide value. Cross-cutting concerns, by their very nature, violate a core design principle – separation of concerns. As a result, OpenTelemetry requires extra care and attention to avoid creating issues for the codebases which depend upon these cross-cutting APIs.
```

> OpenTelemetry is designed to separate the portion of each signal which must be imported as cross-cutting concerns from the portions of OpenTelemetry which can be managed independently. OpenTelemetry is also designed to be a pluggable framework. To accomplish this these goals, each signal consists of four types of packages:

Reviewer 9: > each signal consists of four types of packages

> every signal has its own, independent API package

For C++, we have single source package distribution consisting of api and sdk for all the signals - Thus a single version number for the release. Do we need to revisit it now ?

Author: No, please stick to a single version number if you can. Let me know how I can make it clearer.

There is a reasonable proposal that the SDK may make breaking changes faster than the API, and a separate version number for each would allow us to follow semver in a cleaner fashion. With one number, either the API looks more unstable than it is, or breaking changes to the SDK will be minor version bumps, making the SDK look more stable than it is. So please consider the versioning implications of following semver with a single version number, as we need input on this subject. I would love it if we could stick to one release, one version, but this is a legitimate concern.

Author: Just following up @lalitb. We've come to the conclusion that a single version number is untenable, as it would break semver in some fashion. We expect the SDK to update faster than the API, and have added different long term support grantees for the different components. 

So, please take this into account with how your are packaging your distribution, and future proof it for the event that you may need to make a major version bump to the SDK separately from the API. Also, the minor versions are no longer required to be in sync, so consider if it would be confusing at this point to bump the API minor version if a change was only made to the SDK. Also, please have a look at the support guarantees for the SDK and Contrib from an ABI compatibility standpoint.

Reviewer 9: @tedsuo sure. Regarding versioning, we are discussing this internally, and would come up with a plan during our community meeting next week. We do understand that there are going to more frequent minor version dump in sdk, without any api level changes.
ABI stability challenges are different and tougher to detect in in C++ as compared to other languages. We do provide ABI stability at API level. But it is not provided by the interfaces the SDK provides as implementation hooks to vendor implementers, like exporters, processors, aggregators and propagators. This is something difficult to achieve without running some smart ABI stability detection tools, and not in our plan. The problem right now I see is that once we are bumped to V1.x, we need to increment the major version for any ABI breaking changes. We may need some flexibility here, but let's us discuss this more internally and see the better way to achieve it.

> **API -** API packages consist of the cross-cutting public interfaces used for instrumentation. Any portion of OpenTelemetry which 3rd-party libraries and application code depend upon. To manage different levels of stability, every signal has its own, independent API package. These individual APIs may also be bundled up into a shared global API, for convenience.

Reviewer 21: > every signal has its own, independent API package

Is this a requirement to have separate package per signal? In .NET (and I think other languages), there is one API package and one SDK package, comprising of traces/metrics/context/baggage. We do not produce a package consisting of just one of these signal.

Author: So, the goal is to offer stability in one signal while allowing others to make breaking changes. In most languages, that cannot be done safely if experimental and stable code are in the same package. 

Beyond the various language-specific requirements, like ABI compatibility, I am concerned that mixing experimental and stable code will make it difficult for maintainers and contributors to reason about the effects of the changes they are making, and raises the possibility that an accidental compatibility breakage will occur. A bright line, like separate packages for each signal, makes evaluating changes simpler. 

There is also the issue of end users understanding what is stable and what is not. I do agree that most languages will want to offer a unified API for convenience sake. (I refer to it as a "global api" in this doc, which I admit may not be the best term.) If that API were just a facade for signal-specific APIs, that would make it easier to "hide" experimental signals by leaving them out of the facade, and remind the user that they are experimental by forcing them to require the extra package. 

But again, the goal here is clarity about compatibility requirements. We have to provide an implementation proposal that could reasonably achieve these goals in any language, so that implementors can get a handle on what a reasonable layout and procedure to meet those would look like. 

If .NET needs to deviate in its implementation in order to adhere to language-specific requirements and conventions, that is completely understandable. That said, is it realistic for .NET to provide clarity and compatibility when stable and experimental code is mixed together?

Reviewer 21: Okay - my main question was, once all signals are stable, do we *require* to ship independent package for each of them? i.e instead of OpenTelemetry.API ("global"), should we split and ship OpenTelemetry.API.Metrics, OpenTelemetry.API.Traces, OpenTelemetry.API.Propagators and so on. ? 

OR,
the requirement is only to split stable vs non-stable signals into separate package?
i.e I can have OpenTelemetry.API package consisting of *all* stable signals like Baggage, Context, Traces, and separate packages for each non-stable signal, like OpenTelemetry.API.Metric, OpenTelemetry.API.Logs, each of these shipping as pre-release, or with package names ending in .experimental?

Reviewer 21: > That said, is it realistic for .NET to provide clarity and compatibility when stable and experimental code is mixed together?

The way .NET packaging is done currently is Metrics are part of the same package as Traces. And we marked all Metric methods/classes as "Obsolete", which would sufficiently warn users (compile time warnings) that it is experimental as shown here: https://github.com/open-telemetry/opentelemetry-dotnet/blob/master/src/OpenTelemetry/Metrics/MeterProviderBuilder.cs#L28-L29

(We are still open to splitting metrics into separate package, that's being discussed in .NET sig)

Author: Really, the goal is just ensuring stability and clearly communicating that stability to end users. If that can be done without splitting up the API, I don't see a problem from a versioning and stability perspective. But, would that not create ABI compatibilities? Or is that not an issue in .NET?

Personally, marking beta code as OBSOLETE sounds a hack, but if that is a reasonable practice in .NET I would not take any issue with it. If users are going to respond with a "wtf" I'd suggest taking a different approach.

Reviewer 21: > Personally, marking beta code as OBSOLETE sounds a hack,

Yes. Its can be called hack. But it surely gets people attention, as it generates compile time warnings.
I wish there was an attribute called "Experimental", just like "Obsolete"!

In .NET versioning proposal, we have listed both approaches. Leaving it open to get more feedback from  .NET SIG, but it looks like we are leaning towards separate package for unstable signals. (that aligns more closely with other SIGs as well)

Author: Thank you, really appreciate the thought you are putting into this.

Author: (I also wish our programming languages would do more to explicitly support stability and change management, OBSOLETE is a good start)

**Semantic Conventions -** A schema defining the attributes which describe common concepts and operations which the signal observes. Note that unlike the API or SDK, stable conventions for all signals may be placed in the same package, as they are often useful across different signals.

**Contrib –** plugins and instrumentation that make use of the API or SDK interfaces, but are not part of the core packages necessary for running OTel. The term "contrib" specifically refers to the plugins and instrumentation maintained by the OpenTelemetry organization; it does not refer to third party plugins hosted elsewhere.

## Signal lifecycle

OpenTelemetry is structured around signals. Each signal represents a coherent, stand-alone set of functionality. Each signal follows a lifecycle.

![drawing](img/0143_api_lifecycle.png)

### Lifecycle stages

**Experimental –** Breaking changes and performance issues may occur. Components may not be feature-complete. The experiment may be discarded.

**Stable –** Stability guarantees apply, based on component tpye (API, SDK, Conventions, and Contrib). Long term dependencies may now be taken against these packages.

> **Deprecated –** this signal has been replaced but is still retains the same stability gurantees.

Reviewer 21: ```suggestion
**Deprecated –** this signal has been replaced but is still retains the same stability guarantees.
```

### Deprecation

In theory, signals could be replaced. When this happens, they are marked as deprecated.

Code is only marked as deprecated when the replacement becomes stable. Deprecated code still abides by the same support guarantees as stable code. Deprecated APIs remain stable and backwards compatible.

### Removal

Packages are end-of-life’d by being removed from the release. The release then makes a major version bump.

We currently have no plans for deprecating signals or creating a major version past v1.0.

## Version Numbers

OpenTelemetry follows [semver 2.0](https://semver.org/) conventions, with the following distinction.

For the purposes of versioning, all code within a release is treated as if it were part of a single package. All packages within a release receive the same version number. By keeping version numbers in lockstep, end users can understand which packages belong to which release.

Note that contrib packages are released separately from core packages. Contrib releases are not required to match version numbers with the core releases.

> In some languages, package managers react poorly to experimental packages having a version higher than 0.X. In these cases, a language-specific workaround is required.

Reviewer 13: What are these languages? SemVer 2.0 describes that unstable / pre-release / experimental package versions could be expressed as `vX.Y.Z` then `-rc0` or `-build123`. Even the tools that are not fully SemVer 2.0 compliant, e.g. NuGet package manager - generally respects the experimental packages, where you can label these as `v1.2.3-build333` ... Would you be able to share concrete examples? Is there a plan in these languages to start supporting SemVer 2.0?

Author: Go, specifically. Go maintainers are concerned that experimental packages with versions higher than v0.X will be mishandled by the go module system. 

I think it would be great to have one release version, but use -alpha or similar delimiter for the experimental packages. That would add clarity to users, but I can also see how some package managers could be confused and think those packages are part of a separate release. but if it works, I think it would be helpful to do this.

I would love some clarity on this subject, as I am not experienced with every package manager currently in use by otel.

Reviewer 13: It'd be a bit unnatural for C++ and C# packages to have `-experimental` in the package name. It is also becoming a huge hassle for a project to change the package name in their configuration from `opentelemetry-sdk` to `opentelemetry-sdk-experimental`. Instead of going that route, of having separate package names - I'd rather suggest that we allow for flexibility here.

For example, .NET recommendations for `experimental` or `alpha` package versions from here:
https://docs.microsoft.com/en-us/nuget/create-packages/prerelease-packages

![image](https://user-images.githubusercontent.com/34072974/101839780-d7653780-3af7-11eb-9b9c-b174a41eb2cf.png)

Specifically including the `-alpha` or `-experimental` or `-rc1` as part of **Package Version** rather than as part of package name. This versioning strategy also makes sense for C++ packages and Linux packages. It's also part of SemVer 2.0 guidelines. Splitting from 2 package sets (`sdk`, `api`) into 4 package sets (`sdk`,`sdk-experimental`,`api`,`api-experimental`) is generally NOT what is a common practice for C# and C++ packages, nor what SemVer 2.0 is suggesting.

Can we have provision for some flexibility in this regard? `Go` can do what is within their constraints, whereas for C# and for C++ (and I believe for Java too), it'd make practical sense to include the alpha or pre-release designation as part of Package Version, maintaining the same package name.

From a practical standpoint, I'd see how my customers in Azure cloud are gonna scream loud if I ask them to change the `Package Name` to add `-experimental`, and they'd generally be ok to change the `Package Version` to something that includes `-alpha` or `-rc1` or alike.

Reviewer 6: > Can we have provision for some flexibility in this regard? Go can do what is within their constraints, whereas for C# and for C++

This makes total sense (with the specific case of .Net using `-alpha` in the version rather than in the name), as it complies with what is expected in the .Net world.

In the example, we would want to have a separate `sdk-metrics, version=1.0.0-alpha` package/assembly, more than having `sdk-experimental, version=1.0.0`, for example.

Reviewer 9: I feel for C++ specifically, the term `package` is big vague to describe. There is no centralized packaging system or standard source code package format  (unlike what we have pypi in python, npm for js/nodejs, nuget for C# ... ). We do have third party package managers like vcpkg and canon which we should support on need basis.  Even it is not possible to distribute in binary format, as it is practically not possible to support all platform ( not thinking of ABI for now which is separate discussion).

So the standard distribution would be to release entire source in zip/tar format, which is already done through GitHub release assets. 

We can take advantage of this source distribution, and embed the experimental features/signals behind feature flag ( using standard C `#ifdef` macros ) instead of having a separate package/assembly for these signals. So,(say)  `opentelemetry-sdk v1.0.0` will contain stable version of traces, and experimental logs/metrics behind feature flag. The user then need to explicitly enable these feature flag to use the experimental features.

Reviewer 13: @lalitb - irrespective of what `package` is, I think we can assume that in C++ land we have two dimensions:
- `Package Name`
- `Package Version`

with C++ package name being something like metapackage of `opentelemetry` = `opentelemetry-api` + `opentelemetry-sdk` (whether we actually have to have these split into two packages, good question), and that metapackage would allow a vendor to deliver their own `openelemetry-sdk-${vendor}` --- since we only guarantee ABI stability at API level, and do not guarantee ABI stability for internal SDK interfaces / custom vendor exporters.

My metapoint was that for minor increments and for pre-release packages, we want to adjust the `Package Version` not the `Package Name`. Basically in C++ scenario - `Package Name` is constant, irrelevant whether it is `opentelemetry` as one or `opentelemetry-api` and `opentelemetry-sdk` , plus `opentelemetry-sdk-$vendor`... Basically I want to avoid creating a hassle of maintaining yet another set of `-experimental` packages for us ;-)

Author: @maxgolov I completely agree that package **should not move** when they become stable, or otherwise break beta/rc users in an unnecessary fashion. That would be a terrible experience, and would leave users stranded on the experimental package. I've clarified this in the latest update.

OpenTelemetry is designed to separate the portion of each signal which must be imported as cross-cutting concerns from the portions of OpenTelemetry which can be managed independently. OpenTelemetry is also designed to be an extensible framework. To accomplish this these goals, each signal consists of four types of packages:

**API -** API packages consist of the cross-cutting public interfaces used for instrumentation. Any portion of OpenTelemetry which 3rd-party libraries and application code depend upon. To manage different levels of stability, every signal has its own, independent API package. These individual APIs may also be bundled up into a shared global API, for convenience.

> **SDK -** The implementation of the API. The SDK is managed by the application owner. Note that the SDKs includes additional public interfaces which are not considered part of the API package, as they are not cross-cutting concerns. These public interfaces include constructors, configuration interfaces, and plugin interfaces. Application owners may interact with the SDK; library developers and instrumentation plugins should never directly reverence SDK packages.

Author: Thanks, I clarified this further.

**API Stability -**
No backward-incompatible changes to the API are allowed unless the major version number is incremented. All existing API calls must continue to compile and function against all future minor versions of the same major version. ABI compatibility is offered in languages which require it.

**SDK Stability -**
Public portions of the SDK (constructors, configuration, end-user interfaces) must remain backwards compatible. Internal interfaces are allowed to break; ABI compatibility is not required.

**Semantic Conventions Stability -**
Semantic Conventions may not be removed once they are stable. New conventions may be added to replace usage of older conventions, but the older conventions are never removed, they will only be marked as deprecated in favor of the newer ones.

**Contrib Stability -**
Plugins and instrumentation are kept up to date, and are released simultaneously (or shortly after) the latest release of the API. The goal is to ensure users can update to the latest version of OpenTelemetry, and not be held back by the plugins that they depend on.

> Public portions of contrib packages (constructors, configuration, interfaces) must remain backwards compatible. ABI compatibility not required.

Reviewer 4: Why isn't ABI compatibility required? Do we expect some level of dependency hell to reduce the maintenance burden of contrib a little?

Author: After some discussion, I've changed ABI compatibility into a language-specific concern. It is too difficult to make a broad pronouncement on this nuanced subject.

Telemetry produced by contrib instrumentation must also remain stable and backwards compatible, to avoid breaking alerts and dashboard. This means that existing data may not be mutated or removed without a major version bump. Additional data may be added. This applies to spans, metrics, resources, attributes, events, and any other data types that OpenTelemetry emits.

### Deprecation

In theory, signals could be replaced. When this happens, they are marked as deprecated.

Code is only marked as deprecated when the replacement becomes stable. Deprecated code still abides by the same support guarantees as stable code. Deprecated APIs remain stable and backwards compatible.

### Removal

Packages are end-of-life’d by being removed from the release. The release then makes a major version bump.

We currently have no plans for deprecating signals or creating a major version past v1.0.

> For clarity, it is still possible to create "v2.0" of existing signals without actually moving to v2.0 and breaking support.

Author: thanks, updated.

For example, imagine we develop a new, better tracing API - let's call it AwesomeTrace. We will never mutate the current tracing API into AwesomeTrace. Instead, AwesomeTrace would be added as an entirely new signal which coexists and interoperates with the current tracing signal. This would make adding AwesomeTrace a minor version bump, *not* v2.0. v2.0 would mark the end of support for current tracing, not the addition of AwesomeTrace. And we don't want to ever end that support, if we can help it.

This is not actually a theoretical example. OpenTelemetry already supports two tracing APIs: OpenTelemetry and OpenTracing. We invented a new tracing API, but continue to support the old one.

## Version Numbers

OpenTelemetry follows [semver 2.0](https://semver.org/) conventions, with the following distinction.

OpenTelemetry clients have four components: API, SDK, Semantic Conventions, and Contrib.

For the purposes of versioning, all code within a component is treated as if it were part of a single package, and versioned with the same version number, except for contrib.

> * All API packages version together, across all signals. Signals do not have separate version numbers.

Author: thanks, added

**Major version bump**
Major version bumps only occur when there is a breaking change to a stable interface, or the removal of deprecated signals.

OpenTelemetry values long term support. The expectation is that we will version to v1.0 once the first set of packages are declared stable. OpenTelemetry will then remain at v1.0 for years. There are no plans for a v2.0 of OpenTelemetry at this time. Additional stable packages, such as metrics and logs, will be added as minor version bumps.

**Minor version bump**
Most changes to OpenTelemetry result in a minor version bump.

* New backward-compatible functionality added to any component.
* Breaking changes to internal SDK components.
* Breaking changes to experimental signals.
* New experimental packages are added.
* Experimental packages become stable.

**Patch version bump**
Patch versions make no changes which would require recompilation or potentially break application code. The following are examples of patch fixes.

* Bug fixes which don't require minor version bump per rules above.
* Security fixes.
* Documentation.

> Currently, OpenTelemetry does have plans to backport bug and security fixes to prior minor versions. Security and bug fixes are only applied to the latest minor version. We are committed to making it feasible for end users to stay up to date with the latest version of OpenTelemetry.

Reviewer 5: ```suggestion
Currently, OpenTelemetry does NOT have plans to backport bug and security fixes to prior minor versions. Security and bug fixes are only applied to the latest minor version. We are committed to making it feasible for end users to stay up to date with the latest version of OpenTelemetry.
```

Each signal provides a mechanism for software to describe itself. A codebase, such as an API handler or a database client, takes a dependency on various signals in order to describe itself. OpenTelemetry instrumentation code is then mixed into the other code within that codebase. This makes OpenTelemetry a **cross-cutting concern** - a piece of software which must be mixed into many other pieces of software in order to provide value. Cross-cutting concerns, by their very nature, violate a core design principle – separation of concerns. As a result, OpenTelemetry requires extra care and attention to avoid creating issues for the codebase which depend upon these cross-cutting APIs.

OpenTelemetry is designed to separate the portion of each signal which must be imported as cross-cutting concerns from the portions of OpenTelemetry which can be managed independently. OpenTelemetry is also designed to be an extensible framework. To accomplish this these goals, each signal consists of four types of packages:

> **API -** API packages consist of the cross-cutting public interfaces used for instrumentation. Any portion of OpenTelemetry which 3rd-party libraries and application code depend upon. To manage different levels of stability, every signal has its own, independent API package. These individual APIs may also be bundled up into a shared global API, for convenience.

Reviewer 17: > Any portion of OpenTelemetry which 3rd-party libraries and application code depend upon.

incomplete sentence?

Author: good catch, fixed.

> **SDK -** The implementation of the API. The SDK is managed by the application owner. Note that the SDKs includes additional public interfaces which are not considered part of the API package, as they are not cross-cutting concerns. These public interfaces are defined as **constructors** and **plugin interfaces**. Application owners may interact with SDK constructors; plugin authors may interact with SDK plugin interfaces. Instrumentation authors must never directly reference any SDK package of any kind, only the API.

Reviewer: I don't think plugin (interfaces) are well defined, so it would be good to move the examples from below up here:
> Examples of plugins include the SpanProcessor, Exporter, and Sampler interfaces. Examples of constructors include configuration objects, environment variables, and SDK builders.

Author: good idea, added.

**Contrib –** plugins and instrumentation that make use of the API or SDK interfaces, but are not part of the core packages necessary for running OTel. The term "contrib" specifically refers to the plugins and instrumentation maintained by the OpenTelemetry organization; it does not refer to third party plugins hosted elsewhere.

## Signal lifecycle

OpenTelemetry is structured around signals. Each signal represents a coherent, stand-alone set of functionality. Each signal follows a lifecycle.

![API Lifecycle](img/0143_api_lifecycle.png)

### Lifecycle stages

**Experimental –** Breaking changes and performance issues may occur. Components may not be feature-complete. The experiment may be discarded.

**Stable –** Stability guarantees apply, based on component type (API, SDK, Conventions, and Contrib). Long term dependencies may now be taken against these packages.

**Deprecated –** this signal has been replaced but is still retains the same stability guarantees.

**Removed -** a deprecated signal is no longer supported, and is removed.

> All signal components may become stable together, or one by one in the following order: API, SDK, Semantic Conventions, Contrib.

Reviewer 17: There's no dependency of SemConv or Contrib on SDK, only on API. There may be dependency SemConv ==> Contrib.

Author: I thought about this more. Consider that Contrib can include SDK plugins; and the SDK is configured with resources and may also do resource detection. So, actually the only safe order is: API, SemConv, SDK, Contrib.

We **could** get into more detail and break Contrib apart into API instrumentation and SDK plugins: API, SemConv, API Instrumentation, SDK, SDK Plugins.

The latter option boils down to "the API ecosystem may become stable before the SDK ecosystem." I like that clarity. But so far, we've only talked about Contrib as a single group, not "API Contrib" and "SDK Contrib." Perhaps adding that distinction is worth it?

Author: I went ahead and added the distinction. I think it adds clarity.

When transitioning from experimental to stable to deprecated, packages **should not move or otherwise break how they are imported by users**. Do NOT use and "experimental" directory or package suffix.

Optionally, package **version numbers** MAY include a suffix, such as -alpha, -beta, -rc, or -experimental, to differentiate stable and experimental packages.

### Stability

Once a signal component is marked as stable, the following rules apply until the end of that signal’s existence.

**API Stability -**

> No backward-incompatible changes to the API are allowed unless the major version number is incremented. All existing API calls must continue to compile and function against all future minor versions of the same major version. ABI compatibility for the API may be offered on a language by language basis.

Reviewer 17: > ABI compatibility for the API

don't know what this means

Author: I mean to say that ABI compatibility, as it is defined in each compiled language, MAY be offered by that language, and is recommended. But ABI is not a requirement – each language can decide how critical or useful this form of stability is.

I have found it too difficult to make a sweeping, cross-language statement about ABI compatibility, and whether it is important to maintain for various components. So I'm leaving this up to maintainers for now.

Does that make sense? Let me know how I can clarify further.

ABI stands for [Application Binary Interface](https://en.wikipedia.org/wiki/Application_binary_interface) for anyone wondering.

**SDK Stability -**

> Public portions of the SDK must remain backwards compatible. There are two categories: **plugin interfaces** and **constructors**. Examples of plugin include the SpanProcessor, Exporter, and Sampler interfaces. Examples of constructors include configuration objects, environment variables, and SDK builders.

Reviewer: ```suggestion
Public portions of the SDK must remain backwards compatible. There are two categories: **plugin interfaces** and **constructors**. Examples of plugins include the SpanProcessor, Exporter, and Sampler interfaces. Examples of constructors include configuration objects, environment variables, and SDK builders.
```
I'd move this just below line 32 above (definition of the architecture).

Author: I left it in both places for now; don't feel like it hurts. In the spec we can figure out the best place for this info.

ABI compatibility for SDK plugin interfaces and constructors may be offered on a language by language basis.

**Semantic Conventions Stability -**
Semantic Conventions may not be removed once they are stable. New conventions may be added to replace usage of older conventions, but the older conventions are never removed, they will only be marked as deprecated in favor of the newer ones.

**Contrib Stability -**
Plugins and instrumentation are kept up to date, and are released simultaneously (or shortly after) the latest release of the API. The goal is to ensure users can update to the latest version of OpenTelemetry, and not be held back by the plugins that they depend on.

Public portions of contrib packages (constructors, configuration, interfaces) must remain backwards compatible. ABI compatibility for contrib packages may be offered on a language by language basis.

> Telemetry produced by contrib instrumentation must also remain stable and backwards compatible, to avoid breaking alerts and dashboard. This means that existing data may not be mutated or removed without a major version bump. Additional data may be added. This applies to spans, metrics, resources, attributes, events, and any other data types that OpenTelemetry emits.

Reviewer: ```suggestion
Telemetry produced by contrib instrumentation must also remain stable and backwards compatible, to avoid breaking alerts and dashboards. This means that existing data may not be mutated or removed without a major version bump. Additional data may be added. This applies to spans, metrics, resources, attributes, events, and any other data types that OpenTelemetry emits.
```

### Deprecation

In theory, signals could be replaced. When this happens, they are marked as deprecated.

Code is only marked as deprecated when the replacement becomes stable. Deprecated code still abides by the same support guarantees as stable code. Deprecated APIs remain stable and backwards compatible.

### Removal

Packages are end-of-life’d by being removed from the release. The release then makes a major version bump.

We currently have no plans for deprecating signals or creating a major version past v1.0.

> For clarity, it is still possible to create a new, backwards incompatible version of an existing type of  signal without actually moving to v2.0 and breaking support. Allow me to explain.

Reviewer: ```suggestion
For clarity, it is still possible to create a new, backwards incompatible version of an existing type of signal without actually moving to v2.0 and breaking support. Allow me to explain.
```

Imagine we develop a new, better tracing API - let's call it AwesomeTrace. We will never mutate the current tracing API into AwesomeTrace. Instead, AwesomeTrace would be added as an entirely new signal which coexists and interoperates with the current tracing signal. This would make adding AwesomeTrace a minor version bump, *not* v2.0. v2.0 would mark the end of support for current tracing, not the addition of AwesomeTrace. And we don't want to ever end that support, if we can help it.

This is not actually a theoretical example. OpenTelemetry already supports two tracing APIs: OpenTelemetry and OpenTracing. We invented a new tracing API, but continue to support the old one.

## Version Numbers

OpenTelemetry follows [semver 2.0](https://semver.org/) conventions, with the following distinction.

OpenTelemetry clients have four components: API, SDK, Semantic Conventions, and Contrib.

> For the purposes of versioning, all code within a component is treated as if it were part of a single package, and versioned with the same version number, except for contrib.

Reviewer 17: ```suggestion
For the purposes of versioning, all code within a component is treated as if it were part of a single package, and versioned with the same version number, except for Contrib, which may be a collection of packages versioned separately.
```

Author: good clarification

* API packages for all signals version together, across all signals. Signals do not have separate version numbers. There is one version number that applies to all signals that are included in the API release that is labeled with that particular version number.
* SDK packages for all signals version together, across all signals. There is one version number that applies to all signals that are included in the SDK release that is labeled with that particular version number.
* Semantic Conventions are a single package with a single version number.
* Each contrib package has it's own version.

> Exception: in some languages, package managers may react poorly to experimental packages having a version higher than 0.X. In these cases, a language-specific workaround is required. So far, Go is the only language which has identified this as an issue.

Reviewer 15: This true for both Ruby and Javascript. Both SIGs will version experimental packages as `0.x.y`. When a signal becomes stable its version will be bumped to match the stable API version.

Author: Awesome, thanks for the input. Updated.

* Bug fixes which don't require minor version bump per rules above.
* Security fixes.
* Documentation.

Currently, OpenTelemetry does NOT have plans to backport bug and security fixes to prior minor versions. Security and bug fixes are only applied to the latest minor version. We are committed to making it feasible for end users to stay up to date with the latest version of OpenTelemetry.

## Long Term Support

![long term support](img/0143_long_term.png)

### API support

Major versions of the API will be supported for a minimum of **three years** after the release of the next major API version. Support covers the following areas.

API stability, as defined above, will be maintained.

A version of the SDK which supports the last major version of the API will continue to be maintained during this period. Bug and security fixes will be backported. Additional feature development is not guaranteed.

Contrib packages available when the API is versioned will continue to be maintained for the duration of this period. Bug and security fixes will be backported. Additional feature development is not guaranteed.

### SDK Support

> SDK stability, as defined above, will be maintained for a minimum of **one year** after after the release of the next major SDK version.

Reviewer 17: What are the practical implications of 3yrs for API and 1yr for the rest? In other words, how useful is API if its corresponding SDK is no longer supported (even for security fixes)?

Author: Here is how I intend it to work:

The three year API support explicitly requires that a working version of the SDK, contrib, and semantic conventions remain available and stable for the duration of the API LTS.

The SDK LTS is meant to cover SDK support when the API hasn't versioned. I expect the SDK to require breaking changes more frequently than the API.

So, as long as we don't break the API, our LTS guarantee is one year. If we break the API, LTS goes up to three years for a version of all components required to maintain the final version of the prior API.

Does that make sense?

> ### Contrib Support

Reviewer 21: I haven't found an official guidance on what goes into contrib, and what doesn't. Is there such guidance? Python recently moved off most instrumentations to -contrib repo. .NET has most instrumentation in the main repo, only few in -contrib repo.
Without such a guidance, this portion could be be confusing.

(this is a separate topic, but since its mentioned several times here, want to make sure there is a guideline for what goes into main repo vs what doesn't)

Author: Good point. This could be reworded as "plugins and instrumentation" for clarity.

Author: I clarified that contrib is defined as plugins and instrumentation which are not required by the spec to be part of the SDK release.

Contrib stability, as defined above, will be maintained for a minimum of **one year** after after the release of the next major version of a contrib package.

## OpenTelemetry GA

> The term “OpenTelemetry GA” refers to the point at which a stable version of both tracing and metrics has been released in at least three languages.

Reviewer 21: just confirming - logs are intentionally left out here, right ?

Author: Correct. GA represents the replacement of OpenTracing and OpenCensus.

This is not actually a theoretical example. OpenTelemetry already supports two tracing APIs: OpenTelemetry and OpenTracing. We invented a new tracing API, but continue to support the old one.

## Version Numbers

OpenTelemetry follows [semver 2.0](https://semver.org/) conventions, with the following distinction.

OpenTelemetry clients have four components: API, SDK, Semantic Conventions, and Contrib.

For the purposes of versioning, all code within a component is treated as if it were part of a single package, and versioned with the same version number, except for contrib.

* API packages for all signals version together, across all signals. Signals do not have separate version numbers. There is one version number that applies to all signals that are included in the API release that is labeled with that particular version number.

> * SDK packages for all signals version together, across all signals. There is one version number that applies to all signals that are included in the SDK release that is labeled with that particular version number.

Reviewer: Should we explicitly point out that these do not necessarily have to match to make it less surprising?

Author: good idea
